{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-04-18T07:03:08.947635Z",
     "start_time": "2025-04-18T07:03:07.036508Z"
    }
   },
   "source": [
    "from pkgutil import read_code\n",
    "\n",
    "from mlcroissant import RecordSet\n",
    "\n",
    "from cryoet_data_portal_croissant.gen import generate_mlcroissant\n",
    "import mlcroissant as mlc\n",
    "import json\n",
    "\n",
    "ds = generate_mlcroissant(dataset_ids=[10000])\n",
    "\n",
    "with open(\"test.json\", \"w\") as f:\n",
    "    f.write(json.dumps(ds[0].to_json(), indent=4, default=str) + '\\n')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found the following 2 warning(s) during the validation:\n",
      "  -  [Metadata(S. pombe cells with defocus)] Property \"http://mlcommons.org/croissant/citeAs\" is recommended, but does not exist.\n",
      "  -  [Metadata(S. pombe cells with defocus)] Property \"https://schema.org/version\" is recommended, but does not exist.\n"
     ]
    }
   ],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T07:04:41.071164Z",
     "start_time": "2025-04-18T07:04:41.066780Z"
    }
   },
   "cell_type": "code",
   "source": [
    "dataset = mlc.Dataset.from_metadata(ds[0])\n",
    "dataset.debug = True\n",
    "records = dataset.records(record_set=\"runs\")\n",
    "\n",
    "for record in records:\n",
    "    print(record)\n"
   ],
   "id": "1ae309df5aa318f3",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:absl:Found the following 2 warning(s) during the validation:\n",
      "  -  [Metadata(S. pombe cells with defocus)] Property \"http://mlcommons.org/croissant/citeAs\" is recommended, but does not exist.\n",
      "  -  [Metadata(S. pombe cells with defocus)] Property \"https://schema.org/version\" is recommended, but does not exist.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'runs/id': 249, 'runs/name': '10000/TS_043'}\n",
      "{'runs/id': 241, 'runs/name': '10000/TS_026'}\n",
      "{'runs/id': 242, 'runs/name': '10000/TS_027'}\n",
      "{'runs/id': 243, 'runs/name': '10000/TS_028'}\n",
      "{'runs/id': 244, 'runs/name': '10000/TS_029'}\n",
      "{'runs/id': 245, 'runs/name': '10000/TS_030'}\n",
      "{'runs/id': 246, 'runs/name': '10000/TS_034'}\n",
      "{'runs/id': 247, 'runs/name': '10000/TS_037'}\n",
      "{'runs/id': 248, 'runs/name': '10000/TS_041'}\n",
      "{'runs/id': 250, 'runs/name': '10000/TS_045'}\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T08:23:00.511440Z",
     "start_time": "2025-04-18T08:23:00.427175Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from linkml_runtime.utils.schemaview import SchemaView\n",
    "\n",
    "sv = SchemaView('schema.yaml')\n",
    "\n",
    "print(dir(sv.all_classes()[\"Run\"]))\n",
    "\n",
    "sv.all_classes()[\"Run\"].attributes['tiltseries']"
   ],
   "id": "c1fb920a1cf8a809",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MissingRequiredField', '__annotations__', '__bool__', '__class__', '__contains__', '__dataclass_fields__', '__dataclass_params__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__iter__', '__le__', '__len__', '__lt__', '__match_args__', '__module__', '__ne__', '__new__', '__post_init__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setitem__', '__sizeof__', '__slotnames__', '__str__', '__subclasshook__', '__weakref__', '_as_dict', '_as_json', '_as_json_dumps', '_as_json_obj', '_class_for', '_class_for_curie', '_class_for_uri', '_default', '_get', '_hide_list', '_idempotent', '_if_missing', '_inherited_slots', '_init_from_dict', '_is_empty', '_items', '_keys', '_normalize_inlined', '_normalize_inlined_as_dict', '_normalize_inlined_as_list', '_normalize_inlined_slot', '_setdefault', '_static_default', '_values', 'abstract', 'aliases', 'all_of', 'alt_descriptions', 'annotations', 'any_of', 'apply_to', 'attributes', 'broad_mappings', 'categories', 'children_are_mutually_disjoint', 'class_class_curie', 'class_class_uri', 'class_model_uri', 'class_name', 'class_uri', 'classification_rules', 'close_mappings', 'comments', 'conforms_to', 'contributors', 'created_by', 'created_on', 'defining_slots', 'definition_uri', 'deprecated', 'deprecated_element_has_exact_replacement', 'deprecated_element_has_possible_replacement', 'description', 'disjoint_with', 'exact_mappings', 'exactly_one_of', 'examples', 'extensions', 'from_schema', 'id_prefixes', 'id_prefixes_are_closed', 'implements', 'imported_from', 'in_language', 'in_subset', 'instantiates', 'is_a', 'keywords', 'last_updated_on', 'local_names', 'mappings', 'mixin', 'mixins', 'modified_by', 'name', 'narrow_mappings', 'none_of', 'notes', 'rank', 'related_mappings', 'represents_relationship', 'rules', 'see_also', 'slot_conditions', 'slot_names_unique', 'slot_usage', 'slots', 'source', 'status', 'string_serialization', 'structured_aliases', 'subclass_of', 'title', 'todos', 'tree_root', 'union_of', 'unique_keys', 'values_from']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "SlotDefinition({\n",
       "  'name': 'tiltseries',\n",
       "  'annotations': {'cascade_delete': Annotation(tag='cascade_delete',\n",
       "                                 value=True,\n",
       "                                 extensions={},\n",
       "                                 annotations={})},\n",
       "  'from_schema': 'cdp-dataset-config',\n",
       "  'inverse': 'Tiltseries.run',\n",
       "  'range': 'Tiltseries',\n",
       "  'multivalued': True\n",
       "})"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T04:47:43.033322Z",
     "start_time": "2025-04-18T04:47:43.023471Z"
    }
   },
   "cell_type": "code",
   "source": [
    "TYPE_MAP = {\n",
    "    \"string\": mlc.DataType.TEXT,\n",
    "    \"int\": mlc.DataType.INTEGER,\n",
    "    \"float\": mlc.DataType.FLOAT,\n",
    "    \"boolean\": mlc.DataType.BOOL,\n",
    "    \"date\": mlc.DataType.DATE,\n",
    "    \"datetime\": mlc.DataType.DATE,\n",
    "    \"url\": mlc.DataType.URL,\n",
    "}\n",
    "\n",
    "import re\n",
    "def camel_to_snake(name):\n",
    "    # Add an underscore before each capital letter and convert to lowercase\n",
    "    return re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()\n",
    "\n",
    "def class_to_recordset(linkml_class):\n",
    "    \"\"\"\n",
    "    Convert a class to a RecordSet.\n",
    "\n",
    "    Args:\n",
    "        sv: SchemaView object.\n",
    "        cls: Class to convert.\n",
    "\n",
    "    Returns:\n",
    "        mlc.RecordSet: The converted RecordSet.\n",
    "    \"\"\"\n",
    "\n",
    "    fields = []\n",
    "\n",
    "    clsname = camel_to_snake(linkml_class.name)\n",
    "\n",
    "    for attr, slot_def in linkml_class.attributes.items():\n",
    "        if slot_def.range in TYPE_MAP:\n",
    "            fld = mlc.Field(\n",
    "                id = f\"{clsname}/{slot_def.name}\",\n",
    "                data_types = [TYPE_MAP[slot_def.range]],\n",
    "                name = slot_def.name,\n",
    "                description = slot_def.description,\n",
    "\n",
    "            )\n",
    "        else:\n",
    "            fld = mlc.Field(\n",
    "                id = f\"{clsname}/{slot_def.name}\",\n",
    "                name = slot_def.name,\n",
    "                description = slot_def.description,\n",
    "                references=mlc.Source(id=f\"{slot_def.range}/id\"),\n",
    "            )\n",
    "\n",
    "        fields.append(fld)\n",
    "\n",
    "    rs = mlc.RecordSet(\n",
    "        id = f\"{clsname}\",\n",
    "        name = f\"{clsname}\",\n",
    "        fields = fields,\n",
    "    )\n",
    "\n",
    "    return rs\n",
    "\n",
    "class_to_recordset(sv.all_classes()[\"Run\"]).to_json()"
   ],
   "id": "da11a3c0d30b12cb",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'@type': 'cr:RecordSet',\n",
       " '@id': 'Run',\n",
       " 'name': 'Run',\n",
       " 'field': [{'@type': 'cr:Field',\n",
       "   '@id': 'Run/alignments',\n",
       "   'name': 'alignments',\n",
       "   'description': 'Tiltseries Alignment',\n",
       "   'references': {'@id': 'Alignment/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/annotations',\n",
       "   'name': 'annotations',\n",
       "   'description': 'Metadata about an annotation for a run',\n",
       "   'references': {'@id': 'Annotation/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/dataset',\n",
       "   'name': 'dataset',\n",
       "   'description': 'Dataset that this run is a part of',\n",
       "   'references': {'@id': 'Dataset/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/frames',\n",
       "   'name': 'frames',\n",
       "   'references': {'@id': 'Frame/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/gain_files',\n",
       "   'name': 'gain_files',\n",
       "   'references': {'@id': 'GainFile/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/frame_acquisition_files',\n",
       "   'name': 'frame_acquisition_files',\n",
       "   'references': {'@id': 'FrameAcquisitionFile/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/per_section_parameters',\n",
       "   'name': 'per_section_parameters',\n",
       "   'description': 'Per section parameters for this run',\n",
       "   'references': {'@id': 'PerSectionParameters/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/tiltseries',\n",
       "   'name': 'tiltseries',\n",
       "   'references': {'@id': 'Tiltseries/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/tomogram_voxel_spacings',\n",
       "   'name': 'tomogram_voxel_spacings',\n",
       "   'description': 'Voxel spacings for a run',\n",
       "   'references': {'@id': 'TomogramVoxelSpacing/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/tomograms',\n",
       "   'name': 'tomograms',\n",
       "   'description': 'Metadata describing a tomogram.',\n",
       "   'references': {'@id': 'Tomogram/id'}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/name',\n",
       "   'name': 'name',\n",
       "   'description': 'Short name for this experiment run',\n",
       "   'dataType': 'sc:Text'},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/s3_prefix',\n",
       "   'name': 's3_prefix',\n",
       "   'description': 'The S3 public bucket path where this run is contained',\n",
       "   'dataType': 'sc:Text'},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'Run/https_prefix',\n",
       "   'name': 'https_prefix',\n",
       "   'description': 'The HTTPS directory path where this run is contained url',\n",
       "   'dataType': 'sc:Text'}]}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 19
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "Field = mlc.Field(\n",
    "    id=\"field\",\n",
    "    name=\"field\",\n",
    "    description=\"field\",\n",
    "    data_types=[mlc.DataType.TEXT],\n",
    "    array_shape=\"(-1,)\",\n",
    ")"
   ],
   "id": "e2825aaa67aa823d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "obj = mlc.FileObject(\n",
    "    id=\"id\",\n",
    "    name=\"name\",\n",
    "    description=\"description\",\n",
    "    content_url=\"https://example.com\",\n",
    "    encoding_formats=[\"image/OME-Zarr\"],\n",
    "    same_as=[\"same_as\"],\n",
    "    contained_in=[\"contained_in\"],\n",
    ")\n",
    "obj."
   ],
   "id": "5eee6d43a2fd9aad"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T08:14:09.812521Z",
     "start_time": "2025-04-18T08:14:09.316489Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import cryoet_data_portal as cdp\n",
    "import hashlib\n",
    "\n",
    "client = cdp.Client()\n",
    "tomos = cdp.Tomogram.find(client, [cdp.Tomogram.run.dataset_id == 10000])\n",
    "jtomos = [tomo.to_dict() for tomo in tomos]\n",
    "\n",
    "with open(\"tomos.json\", \"w\") as f:\n",
    "    f.write(json.dumps(jtomos, indent=4, default=str) + '\\n')\n",
    "\n",
    "with open(\"tomos.json\", \"rb\") as f:\n",
    "    csum = hashlib.file_digest(f, hashlib.sha256)\n",
    "\n",
    "print(csum.hexdigest())\n",
    "\n",
    "\n",
    "cdp.Tomogram."
   ],
   "id": "99cf0d6df4326077",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "234b14620d93934980f6131d41251eec918e86b07738434b1554867c506b8077\n"
     ]
    }
   ],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T08:14:19.275115Z",
     "start_time": "2025-04-18T08:14:19.260902Z"
    }
   },
   "cell_type": "code",
   "source": [
    "obj = mlc.FileObject(\n",
    "    id=\"tomogramsfile\",\n",
    "    name=\"name\",\n",
    "    description=\"description\",\n",
    "    content_url=\"http://0.0.0.0:8000/tomos.json\",\n",
    "    encoding_formats=[\"application/json\"],\n",
    "    sha256 = \"234b14620d93934980f6131d41251eec918e86b07738434b1554867c506b8077\",\n",
    ")\n",
    "\n",
    "rs = mlc.RecordSet(\n",
    "    id = \"tomograms\",\n",
    "    name=\"tomograms\",\n",
    "    description=\"Tomograms\",\n",
    "    key=[\"tomograms/id\"],\n",
    "    fields=[\n",
    "        mlc.Field(\n",
    "            id=\"tomograms/id\",\n",
    "            data_types=[mlc.DataType.INTEGER],\n",
    "            description=\"ID of the tomogram.\",\n",
    "            source=mlc.Source(file_object=\"tomogramsfile\", extract=mlc.Extract(json_path=\"$[*].id\")),\n",
    "        ),\n",
    "        mlc.Field(\n",
    "            id=\"tomograms/name\",\n",
    "            data_types=[mlc.DataType.TEXT],\n",
    "            description=\"Name of the tomogram.\",\n",
    "            source=mlc.Source(file_object=\"tomogramsfile\", extract=mlc.Extract(json_path=\"$[*].s3_omezarr_dir\")),\n",
    "        ),\n",
    "    ],\n",
    ")\n",
    "\n",
    "md = mlc.Metadata(\n",
    "    name=\"test\",\n",
    "    description=\"test\",\n",
    "    creators=[mlc.Person(name=\"Test Person\")],\n",
    "    license=[\"https://creativecommons.org/public-domain/cc0/\"],\n",
    "    url=\"https://example.com\",\n",
    "    distribution=[obj],\n",
    "    record_sets=[rs],\n",
    ")\n",
    "\n",
    "ds = mlc.Dataset.from_metadata(md)\n",
    "\n",
    "for record in ds.records(record_set=\"tomograms\"):\n",
    "    print(record)\n"
   ],
   "id": "83a4e54c6134a9c3",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:absl:Found the following 3 warning(s) during the validation:\n",
      "  -  [Metadata(test)] Property \"http://mlcommons.org/croissant/citeAs\" is recommended, but does not exist.\n",
      "  -  [Metadata(test)] Property \"https://schema.org/datePublished\" is recommended, but does not exist.\n",
      "  -  [Metadata(test)] Property \"https://schema.org/version\" is recommended, but does not exist.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'tomograms/id': 629, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_043/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_043.zarr'}\n",
      "{'tomograms/id': 621, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_026/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_026.zarr'}\n",
      "{'tomograms/id': 622, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_027/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_027.zarr'}\n",
      "{'tomograms/id': 623, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_028/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_028.zarr'}\n",
      "{'tomograms/id': 624, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_029/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_029.zarr'}\n",
      "{'tomograms/id': 625, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_030/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_030.zarr'}\n",
      "{'tomograms/id': 626, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_034/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_034.zarr'}\n",
      "{'tomograms/id': 627, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_037/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_037.zarr'}\n",
      "{'tomograms/id': 628, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_041/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_041.zarr'}\n",
      "{'tomograms/id': 630, 'tomograms/name': b's3://cryoet-data-portal-public/10000/TS_045/Reconstructions/VoxelSpacing13.480/Tomograms/100/TS_045.zarr'}\n"
     ]
    }
   ],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "\n",
   "id": "59c7f7542897dc67"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T08:26:50.929097Z",
     "start_time": "2025-04-18T08:26:50.926231Z"
    }
   },
   "cell_type": "code",
   "source": "cdp.Dataset.__doc__",
   "id": "a0461cdc9aec1d8e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A collection of imaging experiments on the same organism\\n\\n    Attributes:\\n        id (int): An identifier for a CryoET dataset, assigned by the Data Portal. Used to identify the dataset as the directory name in data tree\\n        deposition (Deposition): The deposition this dataset is a part of\\n        deposition_id (int): None\\n        funding_sources (List[DatasetFunding]): The dataset fundings of this dataset\\n        authors (List[DatasetAuthor]): The dataset authors of this dataset\\n        runs (List[Run]): The runs of this dataset\\n        title (str): Title of a CryoET dataset\\n        description (str): A short description of a CryoET dataset, similar to an abstract for a journal article or dataset.\\n        organism_name (str): Name of the organism from which a biological sample used in a CryoET study is derived from, e.g. homo sapiens\\n        organism_taxid (int): NCBI taxonomy identifier for the organism, e.g. 9606\\n        tissue_name (str): Name of the tissue from which a biological sample used in a CryoET study is derived from.\\n        tissue_id (str): UBERON identifier for the tissue\\n        cell_name (str): Name of the cell from which a biological sample used in a CryoET study is derived from.\\n        cell_type_id (str): Cell Ontology identifier for the cell type\\n        cell_strain_name (str): Cell line or strain for the sample.\\n        cell_strain_id (str): Link to more information about the cell strain\\n        sample_type (str): Type of samples used in a CryoET study. (cell, tissue, organism, intact organelle, in-vitro mixture, in-silico synthetic data, other)\\n        sample_preparation (str): Describe how the sample was prepared.\\n        grid_preparation (str): Describe Cryo-ET grid preparation.\\n        other_setup (str): Describe other setup not covered by sample preparation or grid preparation that may make this dataset unique in the same publication\\n        key_photo_url (str): URL for the dataset preview image.\\n        key_photo_thumbnail_url (str): URL for the thumbnail of preview image.\\n        cell_component_name (str): Name of the cellular component\\n        cell_component_id (str): If the dataset focuses on a specific part of a cell, the subset is included here\\n        deposition_date (date): Date when a dataset is initially received by the Data Portal.\\n        release_date (date): Date when a dataset is made available on the Data Portal.\\n        last_modified_date (date): Date when a released dataset is last modified.\\n        dataset_publications (str): Comma-separated list of DOIs for publications associated with the dataset.\\n        related_database_entries (str): If a CryoET dataset is also deposited into another database, enter the database identifier here (e.g. EMPIAR-11445). Use a comma to separate multiple identifiers.\\n        s3_prefix (str): The S3 public bucket path where this dataset is contained\\n        https_prefix (str): The https directory path where this dataset is contained\\n        file_size (float): Size of the dataset in bytes\\n    '"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 35
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T09:28:01.929023Z",
     "start_time": "2025-04-18T09:28:01.810814Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from griffe import Docstring\n",
    "from typing import Type\n",
    "from pydantic import BaseModel, create_model\n",
    "\n",
    "_PORTAL_TYPES = (\n",
    "      Type[cdp.Annotation]\n",
    "    | Type[cdp.AnnotationShape]\n",
    "    | Type[cdp.AnnotationFile]\n",
    "    | Type[cdp.Tomogram]\n",
    "    | Type[cdp.TiltSeries]\n",
    "    | Type[cdp.Dataset]\n",
    "    | Type[cdp.Run]\n",
    "    | Type[cdp.Alignment]\n",
    ")\n",
    "\n",
    "_TYPE_MAP = {\n",
    "    \"str\": mlc.DataType.TEXT,\n",
    "    \"int\": mlc.DataType.INTEGER,\n",
    "    \"float\": mlc.DataType.FLOAT,\n",
    "    \"bool\": mlc.DataType.BOOL,\n",
    "    \"date\": mlc.DataType.DATE,\n",
    "    \"datetime\": mlc.DataType.DATE,\n",
    "    \"url\": mlc.DataType.URL,\n",
    "}\n",
    "\n",
    "def get_descriptions(cls: Type[cdp.Dataset]) -> tuple[str, dict[str, str]]:\n",
    "    \"\"\"\n",
    "    Get the description and attributes of a class from its google docstring.\n",
    "\n",
    "    Args:\n",
    "        cls: The class to extract the docstring from.\n",
    "    Returns:\n",
    "        tuple: A tuple containing the description and a dictionary of attributes and their descriptions.\n",
    "    \"\"\"\n",
    "    doc = Docstring(cls.__doc__)\n",
    "    attrs = doc.parse(\"google\")\n",
    "    return attrs[0].value, {a.name: a.description for a in attrs[1].value}\n",
    "\n",
    "def _portal_to_recordset(clz: _PORTAL_TYPES) -> mlc.RecordSet:\n",
    "    \"\"\"Automatically create a Pydantic model from a CryoET Data Portal annotation class.\"\"\"\n",
    "    attrs = clz.__annotations__\n",
    "    docs = get_descriptions(clz)\n",
    "    fields = []\n",
    "    clz_name = clz._gql_type.lower()\n",
    "    filename = f\"http://{clz._gql_root_field}.json\"\n",
    "\n",
    "    for name, typ in attrs.items():\n",
    "        if typ in [\"int\", \"float\", \"str\", \"bool\"] and name[0] != \"_\":\n",
    "            data_type = _TYPE_MAP[typ]\n",
    "            if typ == \"str\" and (\"http\" in name or \"s3://\" in name):\n",
    "                data_type = mlc.DataType.URL\n",
    "\n",
    "            if \"id\" in name and name != \"id\":\n",
    "                fld = mlc.Field(\n",
    "                    id=f\"{clz_name}_{name}\",\n",
    "                    name=name,\n",
    "                    data_types=[data_type],\n",
    "                    description=docs[1].get(name, None),\n",
    "                    source=mlc.Source(\n",
    "                        file_object=filename,\n",
    "                        extract=mlc.Extract(\n",
    "                            json_path=f\"$[*].{name}\"\n",
    "                        ),\n",
    "                    ),\n",
    "                    references=mlc.Source(id=f\"{name}\"),\n",
    "                )\n",
    "            else:\n",
    "                fld = mlc.Field(\n",
    "                    id=f\"{clz_name}_{name}\",\n",
    "                    name=name,\n",
    "                    data_types=[data_type],\n",
    "                    description=docs[1].get(name, None),\n",
    "                    source=mlc.Source(\n",
    "                        file_object=filename,\n",
    "                        extract=mlc.Extract(\n",
    "                            json_path=f\"$[*].{name}\"\n",
    "                        ),\n",
    "                    ),\n",
    "                )\n",
    "\n",
    "            fields.append(fld)\n",
    "\n",
    "    rs = mlc.RecordSet(\n",
    "        id=f\"{clz_name}\",\n",
    "        name=f\"{clz_name}\",\n",
    "        description=docs[0],\n",
    "        fields=fields,\n",
    "        key=[f\"{clz_name}_id\"],\n",
    "    )\n",
    "    return rs\n",
    "\n",
    "rss = _portal_to_recordset(cdp.Dataset)\n",
    "\n",
    "rss.to_json()"
   ],
   "id": "306e1430a954450b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'@type': 'cr:RecordSet',\n",
       " '@id': 'dataset',\n",
       " 'name': 'dataset',\n",
       " 'description': 'A collection of imaging experiments on the same organism',\n",
       " 'key': {'@id': 'dataset_id'},\n",
       " 'field': [{'@type': 'cr:Field',\n",
       "   '@id': 'dataset_id',\n",
       "   'name': 'id',\n",
       "   'description': 'An identifier for a CryoET dataset, assigned by the Data Portal. Used to identify the dataset as the directory name in data tree',\n",
       "   'dataType': 'sc:Integer',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_deposition_id',\n",
       "   'name': 'deposition_id',\n",
       "   'description': 'None',\n",
       "   'dataType': 'sc:Integer',\n",
       "   'references': {'@id': 'deposition_id'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].deposition_id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_title',\n",
       "   'name': 'title',\n",
       "   'description': 'Title of a CryoET dataset',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].title'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_description',\n",
       "   'name': 'description',\n",
       "   'description': 'A short description of a CryoET dataset, similar to an abstract for a journal article or dataset.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].description'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_organism_name',\n",
       "   'name': 'organism_name',\n",
       "   'description': 'Name of the organism from which a biological sample used in a CryoET study is derived from, e.g. homo sapiens',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].organism_name'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_organism_taxid',\n",
       "   'name': 'organism_taxid',\n",
       "   'description': 'NCBI taxonomy identifier for the organism, e.g. 9606',\n",
       "   'dataType': 'sc:Integer',\n",
       "   'references': {'@id': 'organism_taxid'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].organism_taxid'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_tissue_name',\n",
       "   'name': 'tissue_name',\n",
       "   'description': 'Name of the tissue from which a biological sample used in a CryoET study is derived from.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].tissue_name'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_tissue_id',\n",
       "   'name': 'tissue_id',\n",
       "   'description': 'UBERON identifier for the tissue',\n",
       "   'dataType': 'sc:Text',\n",
       "   'references': {'@id': 'tissue_id'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].tissue_id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_name',\n",
       "   'name': 'cell_name',\n",
       "   'description': 'Name of the cell from which a biological sample used in a CryoET study is derived from.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_name'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_type_id',\n",
       "   'name': 'cell_type_id',\n",
       "   'description': 'Cell Ontology identifier for the cell type',\n",
       "   'dataType': 'sc:Text',\n",
       "   'references': {'@id': 'cell_type_id'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_type_id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_strain_name',\n",
       "   'name': 'cell_strain_name',\n",
       "   'description': 'Cell line or strain for the sample.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_strain_name'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_strain_id',\n",
       "   'name': 'cell_strain_id',\n",
       "   'description': 'Link to more information about the cell strain',\n",
       "   'dataType': 'sc:Text',\n",
       "   'references': {'@id': 'cell_strain_id'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_strain_id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_sample_type',\n",
       "   'name': 'sample_type',\n",
       "   'description': 'Type of samples used in a CryoET study. (cell, tissue, organism, intact organelle, in-vitro mixture, in-silico synthetic data, other)',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].sample_type'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_sample_preparation',\n",
       "   'name': 'sample_preparation',\n",
       "   'description': 'Describe how the sample was prepared.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].sample_preparation'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_grid_preparation',\n",
       "   'name': 'grid_preparation',\n",
       "   'description': 'Describe Cryo-ET grid preparation.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'references': {'@id': 'grid_preparation'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].grid_preparation'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_other_setup',\n",
       "   'name': 'other_setup',\n",
       "   'description': 'Describe other setup not covered by sample preparation or grid preparation that may make this dataset unique in the same publication',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].other_setup'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_key_photo_url',\n",
       "   'name': 'key_photo_url',\n",
       "   'description': 'URL for the dataset preview image.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].key_photo_url'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_key_photo_thumbnail_url',\n",
       "   'name': 'key_photo_thumbnail_url',\n",
       "   'description': 'URL for the thumbnail of preview image.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].key_photo_thumbnail_url'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_component_name',\n",
       "   'name': 'cell_component_name',\n",
       "   'description': 'Name of the cellular component',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_component_name'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_cell_component_id',\n",
       "   'name': 'cell_component_id',\n",
       "   'description': 'If the dataset focuses on a specific part of a cell, the subset is included here',\n",
       "   'dataType': 'sc:Text',\n",
       "   'references': {'@id': 'cell_component_id'},\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].cell_component_id'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_dataset_publications',\n",
       "   'name': 'dataset_publications',\n",
       "   'description': 'Comma-separated list of DOIs for publications associated with the dataset.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].dataset_publications'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_related_database_entries',\n",
       "   'name': 'related_database_entries',\n",
       "   'description': 'If a CryoET dataset is also deposited into another database, enter the database identifier here (e.g. EMPIAR-11445). Use a comma to separate multiple identifiers.',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].related_database_entries'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_s3_prefix',\n",
       "   'name': 's3_prefix',\n",
       "   'description': 'The S3 public bucket path where this dataset is contained',\n",
       "   'dataType': 'sc:Text',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].s3_prefix'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_https_prefix',\n",
       "   'name': 'https_prefix',\n",
       "   'description': 'The https directory path where this dataset is contained',\n",
       "   'dataType': 'sc:URL',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].https_prefix'}}},\n",
       "  {'@type': 'cr:Field',\n",
       "   '@id': 'dataset_file_size',\n",
       "   'name': 'file_size',\n",
       "   'description': 'Size of the dataset in bytes',\n",
       "   'dataType': 'sc:Float',\n",
       "   'source': {'fileObject': {'@id': 'http://datasets.json'},\n",
       "    'extract': {'jsonPath': '$[*].file_size'}}}]}"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 65
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T09:34:48.251972Z",
     "start_time": "2025-04-18T09:34:44.772310Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import cryoet_data_portal as cdp\n",
    "import hashlib\n",
    "\n",
    "def _dump_portal(dataset_id: int) -> tuple[mlc.FileObject, mlc.RecordSet]:\n",
    "    \"\"\"\n",
    "    Dump a CryoET Data Portal class to a list of FileObjects.\n",
    "\n",
    "    Args:\n",
    "        clz: The class to dump.\n",
    "\n",
    "    Returns:\n",
    "        list[mlc.FileObject]: A list of FileObjects.\n",
    "    \"\"\"\n",
    "\n",
    "    to_dump = [\n",
    "        {\n",
    "            \"class\": cdp.Annotation,\n",
    "            \"query\": [cdp.Annotation.run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.AnnotationShape,\n",
    "            \"query\": [cdp.AnnotationShape.annotation.run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.AnnotationFile,\n",
    "            \"query\": [cdp.AnnotationFile.tomogram_voxel_spacing.run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.Tomogram,\n",
    "            \"query\": [cdp.Tomogram.run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.TiltSeries,\n",
    "            \"query\": [cdp.TiltSeries.run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.Dataset,\n",
    "            \"query\": [cdp.Dataset.id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.Run,\n",
    "            \"query\": [cdp.Run.dataset_id == dataset_id]\n",
    "        },\n",
    "        {\n",
    "            \"class\": cdp.Alignment,\n",
    "            \"query\": [cdp.Alignment.run.dataset_id == dataset_id]\n",
    "        },\n",
    "    ]\n",
    "\n",
    "    objects = []\n",
    "    recordsets = []\n",
    "\n",
    "    for dump in to_dump:\n",
    "        clz = dump[\"class\"]\n",
    "        query = dump[\"query\"]\n",
    "        client = cdp.Client()\n",
    "        items = clz.find(client, query)\n",
    "        jitems = [item.to_dict() for item in items]\n",
    "        filename = f\"testdata/{clz._gql_root_field}.json\"\n",
    "        text = json.dumps(jitems, indent=4, default=str) + '\\n'\n",
    "\n",
    "        with open(filename, \"w\") as f:\n",
    "            f.write(text)\n",
    "\n",
    "        with open(filename, \"rb\") as f:\n",
    "            checksum = hashlib.file_digest(f, hashlib.sha256).hexdigest()\n",
    "\n",
    "        print(checksum)\n",
    "\n",
    "        obj = mlc.FileObject(\n",
    "            id=filename,\n",
    "            name=filename,\n",
    "            description=\"description\",\n",
    "            content_url=f\"http://0.0.0.0:8000/{clz._gql_root_field}.json\",\n",
    "            encoding_formats=[\"application/json\"],\n",
    "            sha256 = checksum,\n",
    "        )\n",
    "\n",
    "        objects.append(obj)\n",
    "\n",
    "        recordsets.append(_portal_to_recordset(clz))\n",
    "        print(recordsets[-1].to_json())\n",
    "    return objects, recordsets\n",
    "\n",
    "objs, recordsets = _dump_portal(10000)\n",
    "#rs = _portal_to_recordset(cdp.Dataset)\n",
    "\n",
    "\n",
    "\n",
    "md = mlc.Metadata(\n",
    "    name=\"test\",\n",
    "    description=\"test\",\n",
    "    creators=[mlc.Person(name=\"Test Person\")],\n",
    "    license=[\"https://creativecommons.org/public-domain/cc0/\"],\n",
    "    url=\"https://example.com\",\n",
    "    distribution=objs,\n",
    "    record_sets=recordsets,\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "ds = mlc.Dataset.from_metadata(md)\n",
    "\n",
    "for record in ds.records(record_set=\"tomograms\"):\n",
    "    print(record)"
   ],
   "id": "4e458d5bb5ba8a8c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "e4392e6bbe54bb52dafa0e71a53a90b59fcde7ad28d934d308909bdbaf6bec16\n",
      "{'@type': 'cr:RecordSet', '@id': 'annotation', 'name': 'annotation', 'description': 'Metadata for an annotation', 'key': {'@id': 'annotation_id'}, 'field': [{'@type': 'cr:Field', '@id': 'annotation_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'annotation_run_id', 'name': 'run_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'run_id'}, 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].run_id'}}}, {'@type': 'cr:Field', '@id': 'annotation_deposition_id', 'name': 'deposition_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'deposition_id'}, 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].deposition_id'}}}, {'@type': 'cr:Field', '@id': 'annotation_s3_metadata_path', 'name': 's3_metadata_path', 'description': 'S3 path for the metadata json file for this annotation', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].s3_metadata_path'}}}, {'@type': 'cr:Field', '@id': 'annotation_https_metadata_path', 'name': 'https_metadata_path', 'description': 'HTTPS path for the metadata json file for this annotation', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].https_metadata_path'}}}, {'@type': 'cr:Field', '@id': 'annotation_annotation_publication', 'name': 'annotation_publication', 'description': 'DOIs for publications that describe the dataset. Use a comma to separate multiple DOIs.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].annotation_publication'}}}, {'@type': 'cr:Field', '@id': 'annotation_annotation_method', 'name': 'annotation_method', 'description': 'Describe how the annotation is made (e.g. Manual, crYoLO, Positive Unlabeled Learning, template matching)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].annotation_method'}}}, {'@type': 'cr:Field', '@id': 'annotation_ground_truth_status', 'name': 'ground_truth_status', 'description': 'Whether an annotation is considered ground truth, as determined by the annotator.', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].ground_truth_status'}}}, {'@type': 'cr:Field', '@id': 'annotation_object_id', 'name': 'object_id', 'description': 'Gene Ontology Cellular Component identifier or UniProtKB accession for the annotation object.', 'dataType': 'sc:Text', 'references': {'@id': 'object_id'}, 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].object_id'}}}, {'@type': 'cr:Field', '@id': 'annotation_object_name', 'name': 'object_name', 'description': 'Name of the object being annotated (e.g. ribosome, nuclear pore complex, actin filament, membrane)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].object_name'}}}, {'@type': 'cr:Field', '@id': 'annotation_object_description', 'name': 'object_description', 'description': 'A textual description of the annotation object, can be a longer description to include additional information not covered by the Annotation object name and state.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].object_description'}}}, {'@type': 'cr:Field', '@id': 'annotation_object_state', 'name': 'object_state', 'description': 'Molecule state annotated (e.g. open, closed)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].object_state'}}}, {'@type': 'cr:Field', '@id': 'annotation_object_count', 'name': 'object_count', 'description': 'Number of objects identified', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].object_count'}}}, {'@type': 'cr:Field', '@id': 'annotation_confidence_precision', 'name': 'confidence_precision', 'description': 'Describe the confidence level of the annotation. Precision is defined as the % of annotation objects being true positive', 'dataType': 'sc:Float', 'references': {'@id': 'confidence_precision'}, 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].confidence_precision'}}}, {'@type': 'cr:Field', '@id': 'annotation_confidence_recall', 'name': 'confidence_recall', 'description': 'Describe the confidence level of the annotation. Recall is defined as the % of true positives being annotated correctly', 'dataType': 'sc:Float', 'references': {'@id': 'confidence_recall'}, 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].confidence_recall'}}}, {'@type': 'cr:Field', '@id': 'annotation_ground_truth_used', 'name': 'ground_truth_used', 'description': 'Annotation filename used as ground truth for precision and recall', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].ground_truth_used'}}}, {'@type': 'cr:Field', '@id': 'annotation_annotation_software', 'name': 'annotation_software', 'description': 'Software used for generating this annotation', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].annotation_software'}}}, {'@type': 'cr:Field', '@id': 'annotation_is_curator_recommended', 'name': 'is_curator_recommended', 'description': 'Data curatorâ€™s subjective choice as the best annotation of the same annotation object ID', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].is_curator_recommended'}}}, {'@type': 'cr:Field', '@id': 'annotation_method_type', 'name': 'method_type', 'description': 'The method type for generating the annotation (e.g. manual, hybrid, automated)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotations.json'}, 'extract': {'jsonPath': '$[*].method_type'}}}]}\n",
      "5e053141fee0eecb777c2b6752649df3501ad9b55abaf1de3f08c2b081949055\n",
      "{'@type': 'cr:RecordSet', '@id': 'annotationshape', 'name': 'annotationshape', 'description': 'Shapes associated with an annotation', 'key': {'@id': 'annotationshape_id'}, 'field': [{'@type': 'cr:Field', '@id': 'annotationshape_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://annotationShapes.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'annotationshape_annotation_id', 'name': 'annotation_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'annotation_id'}, 'source': {'fileObject': {'@id': 'http://annotationShapes.json'}, 'extract': {'jsonPath': '$[*].annotation_id'}}}, {'@type': 'cr:Field', '@id': 'annotationshape_shape_type', 'name': 'shape_type', 'description': 'The shape of the annotation (SegmentationMask, OrientedPoint, Point, InstanceSegmentation, Mesh)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotationShapes.json'}, 'extract': {'jsonPath': '$[*].shape_type'}}}]}\n",
      "fbfe246fd0749155dc616ddfed7f9a7e63f4dbe3ef4e53ca08c3761d98a753a5\n",
      "{'@type': 'cr:RecordSet', '@id': 'annotationfile', 'name': 'annotationfile', 'description': 'Metadata for files associated with an annotation', 'key': {'@id': 'annotationfile_id'}, 'field': [{'@type': 'cr:Field', '@id': 'annotationfile_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_alignment_id', 'name': 'alignment_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'alignment_id'}, 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].alignment_id'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_annotation_shape_id', 'name': 'annotation_shape_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'annotation_shape_id'}, 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].annotation_shape_id'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_tomogram_voxel_spacing_id', 'name': 'tomogram_voxel_spacing_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'tomogram_voxel_spacing_id'}, 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].tomogram_voxel_spacing_id'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_format', 'name': 'format', 'description': 'File format for this file', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].format'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_s3_path', 'name': 's3_path', 'description': 's3 path of the annotation file', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].s3_path'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_file_size', 'name': 'file_size', 'description': 'Size of annota file in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].file_size'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_https_path', 'name': 'https_path', 'description': 'HTTPS path for this annotation file', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].https_path'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_is_visualization_default', 'name': 'is_visualization_default', 'description': 'Data curatorâ€™s subjective choice of default annotation to display in visualization for an object', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].is_visualization_default'}}}, {'@type': 'cr:Field', '@id': 'annotationfile_source', 'name': 'source', 'description': 'The source type for the annotation file (dataset_author, community, or portal_standard)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://annotationFiles.json'}, 'extract': {'jsonPath': '$[*].source'}}}]}\n",
      "234b14620d93934980f6131d41251eec918e86b07738434b1554867c506b8077\n",
      "{'@type': 'cr:RecordSet', '@id': 'tomogram', 'name': 'tomogram', 'description': 'Metadata describing a tomogram.', 'key': {'@id': 'tomogram_id'}, 'field': [{'@type': 'cr:Field', '@id': 'tomogram_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'tomogram_alignment_id', 'name': 'alignment_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'alignment_id'}, 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].alignment_id'}}}, {'@type': 'cr:Field', '@id': 'tomogram_deposition_id', 'name': 'deposition_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'deposition_id'}, 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].deposition_id'}}}, {'@type': 'cr:Field', '@id': 'tomogram_run_id', 'name': 'run_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'run_id'}, 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].run_id'}}}, {'@type': 'cr:Field', '@id': 'tomogram_tomogram_voxel_spacing_id', 'name': 'tomogram_voxel_spacing_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'tomogram_voxel_spacing_id'}, 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].tomogram_voxel_spacing_id'}}}, {'@type': 'cr:Field', '@id': 'tomogram_name', 'name': 'name', 'description': 'Short name for this tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].name'}}}, {'@type': 'cr:Field', '@id': 'tomogram_size_x', 'name': 'size_x', 'description': 'Number of pixels in the 3D data fast axis', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].size_x'}}}, {'@type': 'cr:Field', '@id': 'tomogram_size_y', 'name': 'size_y', 'description': 'Number of pixels in the 3D data medium axis', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].size_y'}}}, {'@type': 'cr:Field', '@id': 'tomogram_size_z', 'name': 'size_z', 'description': 'Number of pixels in the 3D data slow axis.  This is the image projection direction at zero stage tilt', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].size_z'}}}, {'@type': 'cr:Field', '@id': 'tomogram_voxel_spacing', 'name': 'voxel_spacing', 'description': 'Voxel spacing equal in all three axes in angstroms', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].voxel_spacing'}}}, {'@type': 'cr:Field', '@id': 'tomogram_fiducial_alignment_status', 'name': 'fiducial_alignment_status', 'description': 'Fiducial Alignment status: True = aligned with fiducial False = aligned without fiducial', 'dataType': 'sc:Text', 'references': {'@id': 'fiducial_alignment_status'}, 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].fiducial_alignment_status'}}}, {'@type': 'cr:Field', '@id': 'tomogram_reconstruction_method', 'name': 'reconstruction_method', 'description': 'Describe reconstruction method (WBP, SART, SIRT)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].reconstruction_method'}}}, {'@type': 'cr:Field', '@id': 'tomogram_processing', 'name': 'processing', 'description': 'Describe additional processing used to derive the tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].processing'}}}, {'@type': 'cr:Field', '@id': 'tomogram_tomogram_version', 'name': 'tomogram_version', 'description': 'Version of tomogram', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].tomogram_version'}}}, {'@type': 'cr:Field', '@id': 'tomogram_processing_software', 'name': 'processing_software', 'description': 'Processing software used to derive the tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].processing_software'}}}, {'@type': 'cr:Field', '@id': 'tomogram_reconstruction_software', 'name': 'reconstruction_software', 'description': 'Name of software used for reconstruction', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].reconstruction_software'}}}, {'@type': 'cr:Field', '@id': 'tomogram_is_portal_standard', 'name': 'is_portal_standard', 'description': 'whether this tomogram adheres to portal standards', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].is_portal_standard'}}}, {'@type': 'cr:Field', '@id': 'tomogram_is_author_submitted', 'name': 'is_author_submitted', 'description': 'Whether this tomogram was submitted by the author of the dataset it belongs to.', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].is_author_submitted'}}}, {'@type': 'cr:Field', '@id': 'tomogram_is_visualization_default', 'name': 'is_visualization_default', 'description': 'Data curatorâ€™s subjective choice of default tomogram to display in visualization for a run', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].is_visualization_default'}}}, {'@type': 'cr:Field', '@id': 'tomogram_s3_omezarr_dir', 'name': 's3_omezarr_dir', 'description': 'S3 path to this tomogram in multiscale OME-Zarr format', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].s3_omezarr_dir'}}}, {'@type': 'cr:Field', '@id': 'tomogram_https_omezarr_dir', 'name': 'https_omezarr_dir', 'description': 'HTTPS path to this tomogram in multiscale OME-Zarr format', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].https_omezarr_dir'}}}, {'@type': 'cr:Field', '@id': 'tomogram_file_size_omezarr', 'name': 'file_size_omezarr', 'description': 'Size of the tomogram in OME-Zarr format in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].file_size_omezarr'}}}, {'@type': 'cr:Field', '@id': 'tomogram_s3_mrc_file', 'name': 's3_mrc_file', 'description': 'S3 path to this tomogram in MRC format (no scaling)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].s3_mrc_file'}}}, {'@type': 'cr:Field', '@id': 'tomogram_https_mrc_file', 'name': 'https_mrc_file', 'description': 'HTTPS path to this tomogram in MRC format (no scaling)', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].https_mrc_file'}}}, {'@type': 'cr:Field', '@id': 'tomogram_file_size_mrc', 'name': 'file_size_mrc', 'description': 'Size of the tomogram in MRC format in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].file_size_mrc'}}}, {'@type': 'cr:Field', '@id': 'tomogram_scale_0_dimensions', 'name': 'scale_0_dimensions', 'description': 'comma separated x,y,z dimensions of the unscaled tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].scale_0_dimensions'}}}, {'@type': 'cr:Field', '@id': 'tomogram_scale_1_dimensions', 'name': 'scale_1_dimensions', 'description': 'comma separated x,y,z dimensions of the scale1 tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].scale_1_dimensions'}}}, {'@type': 'cr:Field', '@id': 'tomogram_scale_2_dimensions', 'name': 'scale_2_dimensions', 'description': 'comma separated x,y,z dimensions of the scale2 tomogram', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].scale_2_dimensions'}}}, {'@type': 'cr:Field', '@id': 'tomogram_ctf_corrected', 'name': 'ctf_corrected', 'description': 'Whether this tomogram is CTF corrected', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].ctf_corrected'}}}, {'@type': 'cr:Field', '@id': 'tomogram_offset_x', 'name': 'offset_x', 'description': 'x offset data relative to the canonical tomogram in pixels', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].offset_x'}}}, {'@type': 'cr:Field', '@id': 'tomogram_offset_y', 'name': 'offset_y', 'description': 'y offset data relative to the canonical tomogram in pixels', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].offset_y'}}}, {'@type': 'cr:Field', '@id': 'tomogram_offset_z', 'name': 'offset_z', 'description': 'z offset data relative to the canonical tomogram in pixels', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].offset_z'}}}, {'@type': 'cr:Field', '@id': 'tomogram_key_photo_url', 'name': 'key_photo_url', 'description': 'URL for the key photo', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].key_photo_url'}}}, {'@type': 'cr:Field', '@id': 'tomogram_key_photo_thumbnail_url', 'name': 'key_photo_thumbnail_url', 'description': 'URL for the thumbnail of key photo', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].key_photo_thumbnail_url'}}}, {'@type': 'cr:Field', '@id': 'tomogram_neuroglancer_config', 'name': 'neuroglancer_config', 'description': 'the compact json of neuroglancer config', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].neuroglancer_config'}}}, {'@type': 'cr:Field', '@id': 'tomogram_publications', 'name': 'publications', 'description': 'Comma-separated list of DOIs for publications associated with the tomogram.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].publications'}}}, {'@type': 'cr:Field', '@id': 'tomogram_related_database_entries', 'name': 'related_database_entries', 'description': 'If a CryoET tomogram is also deposited into another database, enter the database identifier here (e.g. EMPIAR-11445). Use a comma to separate multiple identifiers.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tomograms.json'}, 'extract': {'jsonPath': '$[*].related_database_entries'}}}]}\n",
      "c2f6d9f8c851766935630dbfc749ab0f94c8dca6974127328047d743ce2d699c\n",
      "{'@type': 'cr:RecordSet', '@id': 'tiltseries', 'name': 'tiltseries', 'description': 'None', 'key': {'@id': 'tiltseries_id'}, 'field': [{'@type': 'cr:Field', '@id': 'tiltseries_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_run_id', 'name': 'run_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'run_id'}, 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].run_id'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_deposition_id', 'name': 'deposition_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'deposition_id'}, 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].deposition_id'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_s3_omezarr_dir', 'name': 's3_omezarr_dir', 'description': 'S3 path to this tiltseries in multiscale OME-Zarr format', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].s3_omezarr_dir'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_file_size_omezarr', 'name': 'file_size_omezarr', 'description': 'Size of the tiltseries in OME-Zarr format in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].file_size_omezarr'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_s3_mrc_file', 'name': 's3_mrc_file', 'description': 'S3 path to this tiltseries in MRC format (no scaling)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].s3_mrc_file'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_file_size_mrc', 'name': 'file_size_mrc', 'description': 'Size of the tiltseries in MRC format in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].file_size_mrc'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_https_omezarr_dir', 'name': 'https_omezarr_dir', 'description': 'HTTPS path to this tiltseries in multiscale OME-Zarr format', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].https_omezarr_dir'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_https_mrc_file', 'name': 'https_mrc_file', 'description': 'HTTPS path to this tiltseries in MRC format (no scaling)', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].https_mrc_file'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_s3_angle_list', 'name': 's3_angle_list', 'description': 'S3 path to the angle list file for this tiltseries', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].s3_angle_list'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_https_angle_list', 'name': 'https_angle_list', 'description': 'HTTPS path to the angle list file for this tiltseries', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].https_angle_list'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_acceleration_voltage', 'name': 'acceleration_voltage', 'description': 'Electron Microscope Accelerator voltage in volts', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].acceleration_voltage'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_spherical_aberration_constant', 'name': 'spherical_aberration_constant', 'description': 'Spherical Aberration Constant of the objective lens in millimeters', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].spherical_aberration_constant'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_manufacturer', 'name': 'microscope_manufacturer', 'description': 'Name of the microscope manufacturer (FEI, TFS, JEOL)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_manufacturer'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_model', 'name': 'microscope_model', 'description': 'Microscope model name', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_model'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_energy_filter', 'name': 'microscope_energy_filter', 'description': 'Energy filter setup used', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_energy_filter'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_phase_plate', 'name': 'microscope_phase_plate', 'description': 'Phase plate configuration', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_phase_plate'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_image_corrector', 'name': 'microscope_image_corrector', 'description': 'Image corrector setup', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_image_corrector'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_microscope_additional_info', 'name': 'microscope_additional_info', 'description': 'Other microscope optical setup information, in addition to energy filter, phase plate and image corrector', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].microscope_additional_info'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_camera_manufacturer', 'name': 'camera_manufacturer', 'description': 'Name of the camera manufacturer', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].camera_manufacturer'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_camera_model', 'name': 'camera_model', 'description': 'Camera model name', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].camera_model'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_min', 'name': 'tilt_min', 'description': 'Minimal tilt angle in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_min'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_max', 'name': 'tilt_max', 'description': 'Maximal tilt angle in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_max'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_range', 'name': 'tilt_range', 'description': 'Total tilt range in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_range'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_step', 'name': 'tilt_step', 'description': 'Tilt step in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_step'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilting_scheme', 'name': 'tilting_scheme', 'description': 'The order of stage tilting during acquisition of the data', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilting_scheme'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_axis', 'name': 'tilt_axis', 'description': 'Rotation angle in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_axis'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_total_flux', 'name': 'total_flux', 'description': 'Number of Electrons reaching the specimen in a square Angstrom area for the entire tilt series', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].total_flux'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_data_acquisition_software', 'name': 'data_acquisition_software', 'description': 'Software used to collect data', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].data_acquisition_software'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_related_empiar_entry', 'name': 'related_empiar_entry', 'description': 'If a tilt series is deposited into EMPIAR, enter the EMPIAR dataset identifier', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].related_empiar_entry'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_binning_from_frames', 'name': 'binning_from_frames', 'description': 'Describes the binning factor from frames to tilt series file', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].binning_from_frames'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_tilt_series_quality', 'name': 'tilt_series_quality', 'description': 'Author assessment of tilt series quality within the dataset (1-5, 5 is best)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].tilt_series_quality'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_is_aligned', 'name': 'is_aligned', 'description': 'Whether this tilt series is aligned', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].is_aligned'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_pixel_spacing', 'name': 'pixel_spacing', 'description': 'Pixel spacing equal in both axes in angstroms', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].pixel_spacing'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_aligned_tiltseries_binning', 'name': 'aligned_tiltseries_binning', 'description': 'Binning factor of the aligned tilt series', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].aligned_tiltseries_binning'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_size_x', 'name': 'size_x', 'description': 'Number of pixels in the 3D data fast axis', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].size_x'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_size_y', 'name': 'size_y', 'description': 'Number of pixels in the 3D data medium axis', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].size_y'}}}, {'@type': 'cr:Field', '@id': 'tiltseries_size_z', 'name': 'size_z', 'description': 'Number of pixels in the 3D data slow axis', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://tiltseries.json'}, 'extract': {'jsonPath': '$[*].size_z'}}}]}\n",
      "176c9dc7a0c20eb5183b251bdc28db60f881bbbf70919291986342df0d9303ad\n",
      "{'@type': 'cr:RecordSet', '@id': 'dataset', 'name': 'dataset', 'description': 'A collection of imaging experiments on the same organism', 'key': {'@id': 'dataset_id'}, 'field': [{'@type': 'cr:Field', '@id': 'dataset_id', 'name': 'id', 'description': 'An identifier for a CryoET dataset, assigned by the Data Portal. Used to identify the dataset as the directory name in data tree', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'dataset_deposition_id', 'name': 'deposition_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'deposition_id'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].deposition_id'}}}, {'@type': 'cr:Field', '@id': 'dataset_title', 'name': 'title', 'description': 'Title of a CryoET dataset', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].title'}}}, {'@type': 'cr:Field', '@id': 'dataset_description', 'name': 'description', 'description': 'A short description of a CryoET dataset, similar to an abstract for a journal article or dataset.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].description'}}}, {'@type': 'cr:Field', '@id': 'dataset_organism_name', 'name': 'organism_name', 'description': 'Name of the organism from which a biological sample used in a CryoET study is derived from, e.g. homo sapiens', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].organism_name'}}}, {'@type': 'cr:Field', '@id': 'dataset_organism_taxid', 'name': 'organism_taxid', 'description': 'NCBI taxonomy identifier for the organism, e.g. 9606', 'dataType': 'sc:Integer', 'references': {'@id': 'organism_taxid'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].organism_taxid'}}}, {'@type': 'cr:Field', '@id': 'dataset_tissue_name', 'name': 'tissue_name', 'description': 'Name of the tissue from which a biological sample used in a CryoET study is derived from.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].tissue_name'}}}, {'@type': 'cr:Field', '@id': 'dataset_tissue_id', 'name': 'tissue_id', 'description': 'UBERON identifier for the tissue', 'dataType': 'sc:Text', 'references': {'@id': 'tissue_id'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].tissue_id'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_name', 'name': 'cell_name', 'description': 'Name of the cell from which a biological sample used in a CryoET study is derived from.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_name'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_type_id', 'name': 'cell_type_id', 'description': 'Cell Ontology identifier for the cell type', 'dataType': 'sc:Text', 'references': {'@id': 'cell_type_id'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_type_id'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_strain_name', 'name': 'cell_strain_name', 'description': 'Cell line or strain for the sample.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_strain_name'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_strain_id', 'name': 'cell_strain_id', 'description': 'Link to more information about the cell strain', 'dataType': 'sc:Text', 'references': {'@id': 'cell_strain_id'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_strain_id'}}}, {'@type': 'cr:Field', '@id': 'dataset_sample_type', 'name': 'sample_type', 'description': 'Type of samples used in a CryoET study. (cell, tissue, organism, intact organelle, in-vitro mixture, in-silico synthetic data, other)', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].sample_type'}}}, {'@type': 'cr:Field', '@id': 'dataset_sample_preparation', 'name': 'sample_preparation', 'description': 'Describe how the sample was prepared.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].sample_preparation'}}}, {'@type': 'cr:Field', '@id': 'dataset_grid_preparation', 'name': 'grid_preparation', 'description': 'Describe Cryo-ET grid preparation.', 'dataType': 'sc:Text', 'references': {'@id': 'grid_preparation'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].grid_preparation'}}}, {'@type': 'cr:Field', '@id': 'dataset_other_setup', 'name': 'other_setup', 'description': 'Describe other setup not covered by sample preparation or grid preparation that may make this dataset unique in the same publication', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].other_setup'}}}, {'@type': 'cr:Field', '@id': 'dataset_key_photo_url', 'name': 'key_photo_url', 'description': 'URL for the dataset preview image.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].key_photo_url'}}}, {'@type': 'cr:Field', '@id': 'dataset_key_photo_thumbnail_url', 'name': 'key_photo_thumbnail_url', 'description': 'URL for the thumbnail of preview image.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].key_photo_thumbnail_url'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_component_name', 'name': 'cell_component_name', 'description': 'Name of the cellular component', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_component_name'}}}, {'@type': 'cr:Field', '@id': 'dataset_cell_component_id', 'name': 'cell_component_id', 'description': 'If the dataset focuses on a specific part of a cell, the subset is included here', 'dataType': 'sc:Text', 'references': {'@id': 'cell_component_id'}, 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].cell_component_id'}}}, {'@type': 'cr:Field', '@id': 'dataset_dataset_publications', 'name': 'dataset_publications', 'description': 'Comma-separated list of DOIs for publications associated with the dataset.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].dataset_publications'}}}, {'@type': 'cr:Field', '@id': 'dataset_related_database_entries', 'name': 'related_database_entries', 'description': 'If a CryoET dataset is also deposited into another database, enter the database identifier here (e.g. EMPIAR-11445). Use a comma to separate multiple identifiers.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].related_database_entries'}}}, {'@type': 'cr:Field', '@id': 'dataset_s3_prefix', 'name': 's3_prefix', 'description': 'The S3 public bucket path where this dataset is contained', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].s3_prefix'}}}, {'@type': 'cr:Field', '@id': 'dataset_https_prefix', 'name': 'https_prefix', 'description': 'The https directory path where this dataset is contained', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].https_prefix'}}}, {'@type': 'cr:Field', '@id': 'dataset_file_size', 'name': 'file_size', 'description': 'Size of the dataset in bytes', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://datasets.json'}, 'extract': {'jsonPath': '$[*].file_size'}}}]}\n",
      "fa5775eb117c1ae536a98466bcda426245bb219408dad3e4ed8acdda00bd9f48\n",
      "{'@type': 'cr:RecordSet', '@id': 'run', 'name': 'run', 'description': 'None', 'key': {'@id': 'run_id'}, 'field': [{'@type': 'cr:Field', '@id': 'run_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://runs.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'run_dataset_id', 'name': 'dataset_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'dataset_id'}, 'source': {'fileObject': {'@id': 'http://runs.json'}, 'extract': {'jsonPath': '$[*].dataset_id'}}}, {'@type': 'cr:Field', '@id': 'run_name', 'name': 'name', 'description': 'Short name for this experiment run', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://runs.json'}, 'extract': {'jsonPath': '$[*].name'}}}, {'@type': 'cr:Field', '@id': 'run_s3_prefix', 'name': 's3_prefix', 'description': 'The S3 public bucket path where this run is contained', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://runs.json'}, 'extract': {'jsonPath': '$[*].s3_prefix'}}}, {'@type': 'cr:Field', '@id': 'run_https_prefix', 'name': 'https_prefix', 'description': 'The HTTPS directory path where this run is contained url', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://runs.json'}, 'extract': {'jsonPath': '$[*].https_prefix'}}}]}\n",
      "5abae47c975e6b81e2562fe076231f497f3d2e2e0748b58aea526dacce93e6c3\n",
      "{'@type': 'cr:RecordSet', '@id': 'alignment', 'name': 'alignment', 'description': 'Tiltseries Alignment', 'key': {'@id': 'alignment_id'}, 'field': [{'@type': 'cr:Field', '@id': 'alignment_id', 'name': 'id', 'description': 'Numeric identifier (May change!)', 'dataType': 'sc:Integer', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].id'}}}, {'@type': 'cr:Field', '@id': 'alignment_deposition_id', 'name': 'deposition_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'deposition_id'}, 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].deposition_id'}}}, {'@type': 'cr:Field', '@id': 'alignment_tiltseries_id', 'name': 'tiltseries_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'tiltseries_id'}, 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].tiltseries_id'}}}, {'@type': 'cr:Field', '@id': 'alignment_run_id', 'name': 'run_id', 'description': 'None', 'dataType': 'sc:Integer', 'references': {'@id': 'run_id'}, 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].run_id'}}}, {'@type': 'cr:Field', '@id': 'alignment_alignment_type', 'name': 'alignment_type', 'description': 'Whether this a LOCAL or GLOBAL alignment', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].alignment_type'}}}, {'@type': 'cr:Field', '@id': 'alignment_alignment_method', 'name': 'alignment_method', 'description': 'The method used to create this alignment', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].alignment_method'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_x_dimension', 'name': 'volume_x_dimension', 'description': 'X dimension of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_x_dimension'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_y_dimension', 'name': 'volume_y_dimension', 'description': 'Y dimension of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_y_dimension'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_z_dimension', 'name': 'volume_z_dimension', 'description': 'Z dimension of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_z_dimension'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_x_offset', 'name': 'volume_x_offset', 'description': 'X shift of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_x_offset'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_y_offset', 'name': 'volume_y_offset', 'description': 'Y shift of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_y_offset'}}}, {'@type': 'cr:Field', '@id': 'alignment_volume_z_offset', 'name': 'volume_z_offset', 'description': 'Z shift of the reconstruction volume in angstrom', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].volume_z_offset'}}}, {'@type': 'cr:Field', '@id': 'alignment_x_rotation_offset', 'name': 'x_rotation_offset', 'description': 'Additional X rotation of the reconstruction volume in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].x_rotation_offset'}}}, {'@type': 'cr:Field', '@id': 'alignment_tilt_offset', 'name': 'tilt_offset', 'description': 'Additional tilt offset in degrees', 'dataType': 'sc:Float', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].tilt_offset'}}}, {'@type': 'cr:Field', '@id': 'alignment_affine_transformation_matrix', 'name': 'affine_transformation_matrix', 'description': 'A placeholder for the affine transformation matrix.', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].affine_transformation_matrix'}}}, {'@type': 'cr:Field', '@id': 'alignment_s3_alignment_metadata', 'name': 's3_alignment_metadata', 'description': 'S3 path to the metadata file for this alignment', 'dataType': 'sc:Text', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].s3_alignment_metadata'}}}, {'@type': 'cr:Field', '@id': 'alignment_https_alignment_metadata', 'name': 'https_alignment_metadata', 'description': 'HTTPS url to the metadata file for this alignment', 'dataType': 'sc:URL', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].https_alignment_metadata'}}}, {'@type': 'cr:Field', '@id': 'alignment_is_portal_standard', 'name': 'is_portal_standard', 'description': 'Whether this is the portal standard alignment', 'dataType': 'sc:Boolean', 'source': {'fileObject': {'@id': 'http://alignments.json'}, 'extract': {'jsonPath': '$[*].is_portal_standard'}}}]}\n"
     ]
    },
    {
     "ename": "ValidationError",
     "evalue": "Found the following 1 error(s) during the validation:\n  -  [Metadata(test) > RecordSet(annotation)] There is a reference to node with UUID \"http://annotations.json\" in node \"annotation\", but this node doesn't exist.",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mValidationError\u001B[39m                           Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[73]\u001B[39m\u001B[32m, line 90\u001B[39m\n\u001B[32m     85\u001B[39m objs, recordsets = _dump_portal(\u001B[32m10000\u001B[39m)\n\u001B[32m     86\u001B[39m \u001B[38;5;66;03m#rs = _portal_to_recordset(cdp.Dataset)\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m90\u001B[39m md = \u001B[43mmlc\u001B[49m\u001B[43m.\u001B[49m\u001B[43mMetadata\u001B[49m\u001B[43m(\u001B[49m\n\u001B[32m     91\u001B[39m \u001B[43m    \u001B[49m\u001B[43mname\u001B[49m\u001B[43m=\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mtest\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[32m     92\u001B[39m \u001B[43m    \u001B[49m\u001B[43mdescription\u001B[49m\u001B[43m=\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mtest\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[32m     93\u001B[39m \u001B[43m    \u001B[49m\u001B[43mcreators\u001B[49m\u001B[43m=\u001B[49m\u001B[43m[\u001B[49m\u001B[43mmlc\u001B[49m\u001B[43m.\u001B[49m\u001B[43mPerson\u001B[49m\u001B[43m(\u001B[49m\u001B[43mname\u001B[49m\u001B[43m=\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mTest Person\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m)\u001B[49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m     94\u001B[39m \u001B[43m    \u001B[49m\u001B[43mlicense\u001B[49m\u001B[43m=\u001B[49m\u001B[43m[\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mhttps://creativecommons.org/public-domain/cc0/\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m     95\u001B[39m \u001B[43m    \u001B[49m\u001B[43murl\u001B[49m\u001B[43m=\u001B[49m\u001B[33;43m\"\u001B[39;49m\u001B[33;43mhttps://example.com\u001B[39;49m\u001B[33;43m\"\u001B[39;49m\u001B[43m,\u001B[49m\n\u001B[32m     96\u001B[39m \u001B[43m    \u001B[49m\u001B[43mdistribution\u001B[49m\u001B[43m=\u001B[49m\u001B[43mobjs\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m     97\u001B[39m \u001B[43m    \u001B[49m\u001B[43mrecord_sets\u001B[49m\u001B[43m=\u001B[49m\u001B[43mrecordsets\u001B[49m\u001B[43m,\u001B[49m\n\u001B[32m     98\u001B[39m \u001B[43m)\u001B[49m\n\u001B[32m    102\u001B[39m ds = mlc.Dataset.from_metadata(md)\n\u001B[32m    104\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m record \u001B[38;5;129;01min\u001B[39;00m ds.records(record_set=\u001B[33m\"\u001B[39m\u001B[33mtomograms\u001B[39m\u001B[33m\"\u001B[39m):\n",
      "\u001B[36mFile \u001B[39m\u001B[32m<string>:45\u001B[39m, in \u001B[36m__init__\u001B[39m\u001B[34m(self, ctx, id, name, parents, jsonld, JSONLD_TYPE, cite_as, creators, date_created, date_modified, date_published, description, keywords, in_language, license, publisher, same_as, sd_licence, url, version, distribution, record_sets, data_collection, data_collection_type, data_collection_missing_data, data_collection_raw_data, data_collection_timeframe, data_imputation_protocol, data_preprocessing_protocol, data_manipulation_protocol, data_annotation_protocol, data_annotation_platform, data_annotation_analysis, annotations_per_item, annotator_demographics, machine_annotation_tools, data_biases, data_use_cases, data_limitations, data_social_impact, personal_sensitive_information, data_release_maintenance_plan)\u001B[39m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/miniforge3/envs/cdpc/lib/python3.12/site-packages/mlcroissant/_src/structure_graph/nodes/metadata.py:339\u001B[39m, in \u001B[36mMetadata.__post_init__\u001B[39m\u001B[34m(self)\u001B[39m\n\u001B[32m    337\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m node \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m.nodes():\n\u001B[32m    338\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m node.ctx.issues.errors:\n\u001B[32m--> \u001B[39m\u001B[32m339\u001B[39m         \u001B[38;5;28;01mraise\u001B[39;00m ValidationError(node.ctx.issues.report())\n\u001B[32m    341\u001B[39m \u001B[38;5;28mself\u001B[39m.ctx.conforms_to = CroissantVersion.from_jsonld(\n\u001B[32m    342\u001B[39m     \u001B[38;5;28mself\u001B[39m.ctx, \u001B[38;5;28mself\u001B[39m.ctx.conforms_to\n\u001B[32m    343\u001B[39m )\n\u001B[32m    345\u001B[39m \u001B[38;5;66;03m# Share the structure graph in the context\u001B[39;00m\n",
      "\u001B[31mValidationError\u001B[39m: Found the following 1 error(s) during the validation:\n  -  [Metadata(test) > RecordSet(annotation)] There is a reference to node with UUID \"http://annotations.json\" in node \"annotation\", but this node doesn't exist."
     ]
    }
   ],
   "execution_count": 73
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
